deepinv/training_utils.py

```diff
@@ -68,285 +46,507 @@ def train(
+        :param tqdm progress_bar: Progress bar.
+        :returns: The current physics operator, the ground truth, the measurement, and the network reconstruction.
+        """
 
-            with torch.no_grad():
-                if plot_measurements and y.shape != x.shape:
-                    y_reshaped = torch.nn.functional.interpolate(y, size=x.shape[2])
-                    if hasattr(physics_cur, "A_adjoint"):
-                        imgs = [y_reshaped, physics_cur.A_adjoint(y), x_net, x]
-                        caption = (
-                            "From top to bottom: input, backprojection, output, target"
+        progress_bar.set_description(f"Epoch {epoch + 1}")
+
+        self.log_dict_iter = {}
+
+        # random permulation of the dataloaders
+        G_perm = np.random.permutation(self.G)
+
+        for g in G_perm:  # for each dataloader
+
+            x, y, physics_cur = self.get_samples(self.train_iterators, g)
+
+            y = y.to(self.device)
+
+            self.optimizer.zero_grad()
+
+            # Run the forward model
+            x_net = self.model_inference(y=y, physics_cur=physics_cur)
+
+            # Backward step
+            self.backward_pass(g=g, x=x, y=y, x_net=x_net)
+
+            # Compute the metrics over the batch
+            _ = self.batch_metric(x=x, x_net=x_net)
+
+            # Log metrics
+            self.log_metrics()
+
+            progress_bar.set_postfix(self.log_dict_iter)
+
+        return physics_cur, x, y, x_net
+
+    def train(self):
+        r"""
+        Train the model.
+
+        It performs the training process, including the setup, the evaluation, the forward and backward passes,
+        and the visualization.
+
+        :returns: The trained model.
+        """
+
+        self.setup_train()
+        for epoch in range(self.epoch_start, self.epochs):
+            self.epoch_eval(epoch)
+
+            self.model.train()
+
+            self.train_iterators = [iter(loader) for loader in self.train_dataloader]
+            batches = len(self.train_dataloader[self.G - 1])
+
+            for i in (progress_bar := tqdm(range(batches), disable=not self.verbose)):
+                physics_cur, x, y, x_net = self.train_step(epoch, progress_bar)
+
+            self.epoch_wandb_vis(epoch, physics_cur, x=x, y=y, x_net=x_net)
+
+            self.loss_history.append(self.total_loss.avg)
+
+            if self.scheduler:
+                self.scheduler.step()
+
+            # Saving the model
+            save_model(
+                epoch,
+                self.model,
+                self.optimizer,
+                self.ckp_interval,
+                self.epochs,
+                self.loss_history,
+                str(self.save_path),
+                eval_psnr=self.eval_psnr if self.perform_eval else None,
+            )
+
+        if self.wandb_vis:
+            wandb.save("model.h5")
+
+        return self.model
+
+    def validate(self, epoch):
```

Comment:
Rename to e.g. `validation_step`? Currently `validate` sounds like it simply performs data validation like [`sklearn.utils.check_X_y`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_X_y.html)